core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -221,6 +221,180 @@ private[spark] class BlockManager(
     new BlockManager.RemoteBlockDownloadFileManager(this)
   private val maxRemoteBlockToMem = conf.get(config.MAX_REMOTE_BLOCK_SIZE_FETCH_TO_MEM)
 
+  /**
+   * @param blockSize the decrypted size of the block
+   */
+  private abstract class BlockStoreUpdater[T](
+      blockSize: Long,
+      blockId: BlockId,
+      level: StorageLevel,
+      classTag: ClassTag[T],
+      tellMaster: Boolean,
+      keepReadLock: Boolean) {
+
+    /**
+     *  Reads the block content into the memory. If the update of the block store is based on a
+     *  temporary file this could lead to loading the whole file into a ChunkedByteBuffer.
+     */
+    protected def readToByteBuffer(): ChunkedByteBuffer
+
+    protected def blockData(): BlockData
+
+    protected def saveToDiskStore(): Unit
+
+    private def saveDeserializedValuesToMemoryStore(inputStream: InputStream): Boolean = {
+      val values = serializerManager.dataDeserializeStream(blockId, inputStream)(classTag)
+      memoryStore.putIteratorAsValues(blockId, values, classTag) match {
+        case Right(_) => true
+        case Left(iter) =>
+          // If putting deserialized values in memory failed, we will put the bytes directly
+          // to disk, so we don't need this iterator and can close it to free resources
+          // earlier.
+          iter.close()
```

Comment:
I don't think the inputStream will get closed here.  It didn't matter so much in the old code, when it was always reading from a byteBuffer, but now this could hold onto to a FileInputStream

Reply:
As I have seen when the values iterator reaches the end then the stream will be closed, so I called the close via `IOUtils.closeQuietly`.

Reply:
Since we're discussing this, it should probably be in a finally block.